src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs

```diff
@@ -0,0 +1,109 @@
+            var workspace = document.Project.Solution.Workspace;
+            var syntaxContext = document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(workspace, semanticModel, position, cancellationToken);
+            if (!syntaxContext.IsAwaitKeywordContext())
+            {
+                return;
+            }
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var syntaxKinds = document.GetRequiredLanguageService<ISyntaxKindsService>();
+            var completionItem = GetCompletionItem(syntaxContext.TargetToken, generator, syntaxKinds, syntaxFacts);
+            context.AddItem(completionItem);
+        }
+
+        public sealed override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = null, CancellationToken cancellationToken = default)
+        {
+            // IsComplexTextEdit is true when we want to add async to the container.
+            if (!item.IsComplexTextEdit)
+            {
+                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var declaration = GetAsyncSupportingDeclaration(root.FindToken(item.Span.Start));
+            if (declaration is null)
+            {
+                // IsComplexTextEdit should only be true when GetAsyncSupportingDeclaration returns non-null.
+                // This is ensured by the ShouldMakeContainerAsync overrides.
+                Debug.Assert(false, "Expected non-null value for declaration.");
+                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+            }
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxKinds = document.GetRequiredLanguageService<ISyntaxKindsService>();
+
+            using var _ = ArrayBuilder<TextChange>.GetInstance(out var builder);
+            builder.Add(new TextChange(new TextSpan(GetSpanStart(declaration), 0), syntaxFacts.GetText(syntaxKinds.AsyncKeyword) + " "));
+            builder.Add(new TextChange(item.Span, item.DisplayText));
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var newText = text.WithChanges(builder);
+            return CompletionChange.Create(Utilities.Collapse(newText, builder.ToImmutableArray()));
+        }
+
+#pragma warning disable VSTHRD200 // Use "Async" suffix for async methods
+        private protected bool ShouldMakeContainerAsync(SyntaxToken token, SyntaxGenerator generator)
+#pragma warning restore VSTHRD200 // Use "Async" suffix for async methods
+        {
+            var declaration = GetAsyncSupportingDeclaration(token);
+            return declaration is not null && !generator.GetModifiers(declaration).IsAsync;
```

Comment:
TODO (just as a reminder to myself): Add tests with existing "async" keyword to hit this code path. I don't trust `SyntaxGenerator` enough 😄

Reply:
Indeed `GetModifiers` was broken. Fixed it.